### DISCUSSION TITLE
Homework 4

### DISCUSSION THREAD

**Discussion by Gobi-Kumarasamy** [cpp-for-yourself/lectures-and-homeworks#46] [Homeworks Q&A] [answered]
Hi,
      While doing pull request, the homework checker throws some error. I am not sure the problem is in my side or not ?
I would like to get some clarity over it. Thank you.

ERROR:
`cp: cannot stat '/home/runner/work/CPP_homeworks/CPP_homeworks/checker/inject/homework_4/validation_tests': No such file or directory`

The screenshot of the homework checker:
![image](https://user-images.githubusercontent.com/67315845/212107932-8dfaac77-6961-4648-8021-758df25289b6.png)

**Comment by niosus** [comment, marked as answer]
Hmm, that's strange. I'm away from my laptop until the weekend but will definitely look at it over the weekend.

**Reply by niosus** [reply]
I think I've found it and it's a bug on my side. I hope to fix it today or tomorrow. It slipped through my testing as my own homeworks repo has exactly the same name as the original one and that's what I'be been testing. I'll have to rename my repository to avoid having these errors.

@Gobi-Kumarasamy  I'll ping you again once it's fixed and you should just rerun the actions on your PR (or create a new dummy PR that will trigger the actions anew).

Thanks again for pointing this out!

**Reply by niosus** [reply]
Although.... I've renamed my repository and it doesn't seem to affect the outcome, so I cannot reproduce your issue for now. Looking into your PR: https://github.com/Gobi-Kumarasamy/CPP_homeworks/pull/7 I see that you faced some issues with the way you commit your work. Could you walk me through this? It might be relevant.

**Reply by Gobi-Kumarasamy** [reply]
@niosus Thank you so much for your quick help.

### Regarding Home work checker bug:
I guess the problem still exists. The latest commit also experiences the same issue.

### Regarding the other issues faced in my commit:
[I am working with windows sub system for Linux through Visual studio code]
Until homework 3, I have not faced any git issues.  I am facing all these issues recently, when I work with homework 4.

- **Issue 1:** fatal: 'origin' does not appear to be a git repository fatal: Could not read from remote repository.
Strangely I did not see any push / fetch remotes for the project.
I manually added the origin `git remote add origin https://github.com/****/CPP_homeworks.git`
This might be a problem in my end, I could have done something wrong accidentally. [Refer Stackoverflow](https://stackoverflow.com/questions/32238616/git-push-fatal-origin-does-not-appear-to-be-a-git-repository-fatal-could-n)

- **Issue 2:** warning: adding embedded git repository.
This issue is working with submodules. 
The homeworks parent directory has .git folder and all the git information.
![image](https://user-images.githubusercontent.com/67315845/212468577-55f1fba2-38cf-4dd1-89e3-e19bb977b3a7.png)

Also, the homework_4/no_strings_attached directory also had the .git folder 
I did this actually by running the following right after I created the project folder structure to add the submodule. 
`git init`
`git submodule add https://github.com/google/googletest.git external/googletest` 
![image](https://user-images.githubusercontent.com/67315845/212468632-da0c660f-5540-4f57-9d82-a7c7a1ec6786.png)

This created the warning of adding embedded git repository. 
Then I deleted the .git folder and then executed this `git rm --cached -f ./homework_4/no_strings_attached`
[Refer Stackoverflow](https://stackoverflow.com/a/72108662)

Now, I realise that I should not have executed `git init` inside homework_4 folder.

Afterwards no issues in git commit and push.

**Reply by niosus** [reply]
@Gobi-Kumarasamy I'm pretty sure the fact that you called `git init` messed up the workflow a bit. But the main issue is on my side. I have updated the CI workflow that is used to check the homeworks at some point to reuse a workflow in the separate repo: https://github.com/cpp-for-yourself/ci-jobs. This new job definition introduces also the copying of the `inject` folder that you're missing, so mostly the issue is on me.

To compensate for your pain, I've looked at your PR and copied all the code into a new repository. I further created a PR that mimics yours with additional 3 commits that undo the issues that you face, please have a look here: https://github.com/niosus/Gobi-Kumarasamy-homeworks/pull/1. In that PR, all the tests pass.

Figuring out what git does is not a trivial task, so I do hope that you will be able to fix your PR after looking at the one I suggest above.

**Reply by Gobi-Kumarasamy** [reply]
Perfect .!
It is passing all the tests.

Thank you so much for the detailed support.! :)

**Reply by niosus** [reply]
Perfect! Glad that it works now and sorry for the hassle!

**Comment by jonasdieker** [comment]
Hi @niosus, first of all thank you for all the great work!

I also seem to have issues with the `inject and build validation tests` part.

At first I had to change one of my header file names because I got an error that it could not find `string_split.h` even though in the homework description this file is called `split_string.h`. So I guess the tests you inject do not match the naming of what is expected in the tests you inject?

I noticed in general the naming of files/target/etc. sometimes being string_xxx and sometimes xxx_string. This can make it a bit harder to follow, just a tip ;)

After changing this I get the following error. So clearly the enum class is the problem, even though `kBoth` should be a member of `Side`.
```
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp: In member function ‘virtual void TrimValidationTest_TrimEmpty_Test::TestBody()’:
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:12:73: error: ‘kBoth’ is not a member of ‘no_strings_attached::Side’
   12 |   const auto trimmed_both_sides_explicit = Trim(test_string, ' ', Side::kBoth);
      |                                                                         ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:13:63: error: ‘kLeft’ is not a member of ‘no_strings_attached::Side’
   13 |   const auto trimmed_left_side = Trim(test_string, ' ', Side::kLeft);
      |                                                               ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:14:64: error: ‘kRight’ is not a member of ‘no_strings_attached::Side’
   14 |   const auto trimmed_right_side = Trim(test_string, ' ', Side::kRight);
      |                                                                ^~~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp: In member function ‘virtual void TrimValidationTest_TrimTrivial_Test::TestBody()’:
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:24:73: error: ‘kBoth’ is not a member of ‘no_strings_attached::Side’
   24 |   const auto trimmed_both_sides_explicit = Trim(test_string, ' ', Side::kBoth);
      |                                                                         ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:25:63: error: ‘kLeft’ is not a member of ‘no_strings_attached::Side’
   25 |   const auto trimmed_left_side = Trim(test_string, ' ', Side::kLeft);
      |                                                               ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:26:64: error: ‘kRight’ is not a member of ‘no_strings_attached::Side’
   26 |   const auto trimmed_right_side = Trim(test_string, ' ', Side::kRight);
      |                                                                ^~~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp: In member function ‘virtual void TrimValidationTest_TrimCorrect_Test::TestBody()’:
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:40:73: error: ‘kBoth’ is not a member of ‘no_strings_attached::Side’
   40 |   const auto trimmed_both_sides_explicit = Trim(test_string, ' ', Side::kBoth);
      |                                                                         ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:41:63: error: ‘kLeft’ is not a member of ‘no_strings_attached::Side’
   41 |   const auto trimmed_left_side = Trim(test_string, ' ', Side::kLeft);
      |                                                               ^~~~~
/tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/validation_tests/test_string_trim.cpp:42:64: error: ‘kRight’ is not a member of ‘no_strings_attached::Side’
   42 |   const auto trimmed_right_side = Trim(test_string, ' ', Side::kRight);
      |                                                                ^~~~~~
gmake[2]: *** [validation_tests/CMakeFiles/validation_tests.dir/build.make:90: validation_tests/CMakeFiles/validation_tests.dir/test_string_trim.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:1235: validation_tests/CMakeFiles/validation_tests.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2

```

and 

```
-- CMAKE_BUILD_TYPE: Release
-- Configuring done (0.1s)
-- Generating done (0.0s)
-- Build files have been written to: /tmp/homework_checker/no_strings_attached_be01f9bd7e67fdf8eb28ee35431a3dbd/build
[  8%] Built target string_split
[ 17%] Built target string_trim
[ 26%] Built target split_strings
[ 34%] Built target gtest
[ 43%] Built target trim_strings
[ 52%] Built target gtest_main
[ 60%] Built target gmock
[ 69%] Built target split_string_test
[ 78%] Built target trim_string_test
[ 86%] Built target gmock_main
[ 91%] Building CXX object validation_tests/CMakeFiles/validation_tests.dir/test_string_split.cpp.o
[ 95%] Building CXX object validation_tests/CMakeFiles/validation_tests.dir/test_string_trim.cpp.o

```

**Reply by jonasdieker** [reply]
Ok, so I moved the enum class to the header file and now it works. Why was this the solution, I don't think I fully understand it yet?

**Reply by niosus** [reply]
All valid points. I am away from my PC (and mostly internet) until Wednesday but will have a proper look then.
